### DISCUSSION TITLE
多节点模型中coal-ppl不发电，不知道是哪里出问题了

### DISCUSSION THREAD

**Discussion by meng25meng** [iiasa/message_ix#846] [Q&A]
在运行全国省份节点模型的时候，我采用excel输入方式，添加的商品有煤电、铀、电力，层级resource、primary、secondary、final，但发现在Excel中把ext-coal添加上去，在gdx文件中没有ext的数据，我想问一下关于price-commdity这个是外生数据？还是模型计算得到的？如何更改这个这个数据？关于成本的有三种投资、运维、可变，都与这个price-commdity有关吗？谢谢诶您的解答 @lumbric 
![F98CE56A567BC619EB060AA1B0C41210](https://github.com/iiasa/message_ix/assets/147787279/5b481711-37bc-4f22-9dbc-7556af8dcbad)
![00FE8D22E21A103087087DCEFACC5F5E](https://github.com/iiasa/message_ix/assets/147787279/ddc80336-585a-47de-a291-44ae52492a49)
@yiyi1991[https://github.com/yiyi1991](url)我还尝试过没有加ext开采技术，然后把commidity都变成electricity但也是没有发电量，不知道该如何去解决

**Comment by meng25meng** [comment]
是不是还有一种可能就是我的需求太小了？轮不到煤电发电？但是全国所有煤电都没发电量，这个应该不可能吧？ @yiyi1991

**Comment by yiyi1991** [comment]
Thank you for moving your questions here in discussion and closing your previous technical issue. 
Your previous question was "why coal power technologies show zeros in their ACT". 
I assume that what you are trying to do here is to rebuild a toy model with multiple technologies. 

As suggested earlier, probably first try checking the `input` and `output` tabs?
 - do all sub_ppl_* (or some other ways you name your coal power plants) have electricity as their output? 
 - do other gas/nuclear power plants have electricity as their output and do these techs also show zeros in their ACT?
And similar checks that will help you make sure that technologies and commodities have been well-added.

 (My current suspicion is that all `demand` is using the commodity electricity directly, but the links from coal/gas/uranium to electricity are somehow broken. If you see non-zero in ACT in other technologies, then all links are well-built and the model is working well, just with other technologies more attractive than coals. Maybe take a look at how you add your constraints.)

Then consider checking region, price, and extraction **AFTER** you have checked input and output, because you should have ACT levels in your technologies even without them (as shown in blocks 1-3 [here](https://github.com/iiasa/message_ix/blob/main/tutorial/westeros/westeros_fossil_resource.ipynb)). 

What is your ultimate purpose for this MESSAGE run? 
(Both PRICE_COMMODITY and EXT are variables instead of parameters. Regarding your question, maybe what you need to add is `resource_cost` and related constraints, see [here](https://github.com/iiasa/message_ix/blob/main/tutorial/westeros/westeros_share_constraint.ipynb)).

**Reply by meng25meng** [reply]
> Thank you for moving your questions here in discussion and closing your previous technical issue. Your previous question was "why coal power technologies show zeros in their ACT". I assume that what you are trying to do here is to rebuild a toy model with multiple technologies.
> 
> As suggested earlier, probably first try checking the `input` and `output` tabs?
> 
> * do all sub_ppl_* (or some other ways you name your coal power plants) have electricity as their output?
> * do other gas/nuclear power plants have electricity as their output and do these techs also show zeros in their ACT?
>   And similar checks that will help you make sure that technologies and commodities have been well-added.
> 
> (My current suspicion is that all `demand` is using the commodity electricity directly, but the links from coal/gas/uranium to electricity are somehow broken. If you see non-zero in ACT in other technologies, then all links are well-built and the model is working well, just with other technologies more attractive than coals. Maybe take a look at how you add your constraints.)
> 
> Then consider checking region, price, and extraction **AFTER** you have checked input and output, because you should have ACT levels in your technologies even without them (as shown in blocks 1-3 [here](https://github.com/iiasa/message_ix/blob/main/tutorial/westeros/westeros_fossil_resource.ipynb)).
> 
> What is your ultimate purpose for this MESSAGE run? (Both PRICE_COMMODITY and EXT are variables instead of parameters. Regarding your question, maybe what you need to add is `resource_cost` and related constraints, see [here](https://github.com/iiasa/message_ix/blob/main/tutorial/westeros/westeros_share_constraint.ipynb)).

1.我的目的是想要得到全国在2030-2050年每个省份的装机和发电量情况。
2.根据您的提示，我首先检查了我的excel的层级，我用GD省份的能源结构，如下：
![0570DB694E095F612E6AB941A4E7A7A6](https://github.com/iiasa/message_ix/assets/147787279/c54bef5f-a908-4fb2-be29-76e05022b787)
他们最终都是final对应的electricity。
3.您的提示中对于resource的表示。我在Excel中进行了输入如下：
![7895956EA2263CD1C7D039018DAC4F73](https://github.com/iiasa/message_ix/assets/147787279/22b9adff-3390-4374-8056-ffe94ae2e9a5)
4.在gdx文件中在全国层面煤和铀开采对应的ACT都为0，这个是因为还没有转换成电量吗？所以为0，还是别的原因，
![6EC08534BF658443469EEAAAB290FA29](https://github.com/iiasa/message_ix/assets/147787279/d3bbc927-c781-4e8e-b16b-95a4bca977a0)
5.在发电对应的技术中有别的能源都有发电：在这里leve和Marginal分别是代表什么？而且在Marginal中还会有EPS出现，
![DA976ED51DFDFDDB4D22A0286C4501ED](https://github.com/iiasa/message_ix/assets/147787279/2bb83594-5240-4a20-ad8d-6dc335c63489)
感谢您的专业解答

**Comment by yiyi1991** [comment]
> 1.我的目的是想要得到全国在2030-2050年每个省份的装机和发电量情况。

I will then suggest getting in contact with researchers who have been working on MESSAGE China for a longer time. Check [this](https://doi.org/10.1021/acs.est.2c08171) with power sectors and [this](https://doi.org/10.1029/2022EF003255) with province-level modeling.  And other researchers in the author lists in related works (e.g., sub-regional coal [here](https://www.sciencedirect.com/science/article/abs/pii/S0360544220327511?via%3Dihub)), to see if there is a more accessible baseline you can use as an intermediate research input to build your scenarios on. 

> 5.在发电对应的技术中有别的能源都有发电：在这里leve和Marginal分别是代表什么？而且在Marginal中还会有EPS出现，

I saw hydro_* you added has non-zero levels in your gdx output, which is great. Your model is well-built and solved. 
Please use the values in `level` (check the definitions of level and marginal [here](https://www.gams.com/latest/docs/UG_Equations.html), also EPS). 

> 2.根据您的提示，我首先检查了我的excel的层级，我用GD省份的能源结构，如下：

It looks good (and since you have some hydropower levels, it should be good). 

> 4.在gdx文件中在全国层面煤和铀开采对应的ACT都为0，这个是因为还没有转换成电量吗？所以为0，还是别的原因，
> 3.您的提示中对于resource的表示。我在Excel中进行了输入如下：

So the problem now is, the technologies that you have given them resource inputs show zero activity? You then suspect that these zeros are caused by technical issues when adding parameters OR just a result of the system preferring hydro?
One way to diagnose is adding a new dummy resource "water" and giving it to hydro power with higher or lower resource costs.

But at the end the 1st point should solve your problem most directly.

**Reply by meng25meng** [reply]
> > 1.我的目的是想要得到全国在2030-2050年每个省份的装机和发电量情况。
> 
> I will then suggest getting in contact with researchers who have been working on MESSAGE China for a longer time. Check [this](https://doi.org/10.1021/acs.est.2c08171) with power sectors and [this](https://doi.org/10.1029/2022EF003255) with province-level modeling. And other researchers in the author lists in related works (e.g., sub-regional coal [here](https://www.sciencedirect.com/science/article/abs/pii/S0360544220327511?via%3Dihub)), to see if there is a more accessible baseline you can use as an intermediate research input to build your scenarios on.
> 
> > 5.在发电对应的技术中有别的能源都有发电：在这里leve和Marginal分别是代表什么？而且在Marginal中还会有EPS出现，
> 
> I saw hydro_* you added has non-zero levels in your gdx output, which is great. Your model is well-built and solved. Please use the values in `level` (check the definitions of level and marginal [here](https://www.gams.com/latest/docs/UG_Equations.html), also EPS).
> 
> > 2.根据您的提示，我首先检查了我的excel的层级，我用GD省份的能源结构，如下：
> 
> It looks good (and since you have some hydropower levels, it should be good).
> 
> > 4.在gdx文件中在全国层面煤和铀开采对应的ACT都为0，这个是因为还没有转换成电量吗？所以为0，还是别的原因，
> > 3.您的提示中对于resource的表示。我在Excel中进行了输入如下：
> 
> So the problem now is, the technologies that you have given them resource inputs show zero activity? You then suspect that these zeros are caused by technical issues when adding parameters OR just a result of the system preferring hydro? One way to diagnose is adding a new dummy resource "water" and giving it to hydro power with higher or lower resource costs.
> 
> But at the end the 1st point should solve your problem most directly.



> > 1.我的目的是想要得到全国在2030-2050年每个省份的装机和发电量情况。
> 
> I will then suggest getting in contact with researchers who have been working on MESSAGE China for a longer time. Check [this](https://doi.org/10.1021/acs.est.2c08171) with power sectors and [this](https://doi.org/10.1029/2022EF003255) with province-level modeling. And other researchers in the author lists in related works (e.g., sub-regional coal [here](https://www.sciencedirect.com/science/article/abs/pii/S0360544220327511?via%3Dihub)), to see if there is a more accessible baseline you can use as an intermediate research input to build your scenarios on.
> 
> > 5.在发电对应的技术中有别的能源都有发电：在这里leve和Marginal分别是代表什么？而且在Marginal中还会有EPS出现，
> 
> I saw hydro_* you added has non-zero levels in your gdx output, which is great. Your model is well-built and solved. Please use the values in `level` (check the definitions of level and marginal [here](https://www.gams.com/latest/docs/UG_Equations.html), also EPS).
> 
> > 2.根据您的提示，我首先检查了我的excel的层级，我用GD省份的能源结构，如下：
> 
> It looks good (and since you have some hydropower levels, it should be good).
> 
> > 4.在gdx文件中在全国层面煤和铀开采对应的ACT都为0，这个是因为还没有转换成电量吗？所以为0，还是别的原因，
> > 3.您的提示中对于resource的表示。我在Excel中进行了输入如下：
> 
> So the problem now is, the technologies that you have given them resource inputs show zero activity? You then suspect that these zeros are caused by technical issues when adding parameters OR just a result of the system preferring hydro? One way to diagnose is adding a new dummy resource "water" and giving it to hydro power with higher or lower resource costs.
> 
> But at the end the 1st point should solve your problem most directly.

感谢您的指点，我发现在gdx文件中每个省份ext也都有数据了在marginal这一列，我在检查我的代码部分，就是关于excel到处就是去参考一下[https://github.com/iiasa/message_ix/blob/main/tutorial/westeros/westeros_report.ipynb](url)这个吗？
2.就是我发现我的其他能源分别设置了bound-new-capoaity-up、bound-total-capacity-lo和bound-total-capacity-lo约束后，CAP都是设定的最小水平，我是需要再加上一个软约束吗？我发现动态约束部分有好多，growth_new_capacity_up和soft_new_capacity_up这个哪个都可以吗？
![4BBD5777B1D085286C746BE1A12944CD](https://github.com/iiasa/message_ix/assets/147787279/a911ee46-6c79-4144-86a7-f36086e41627)

![4BBD5777B1D085286C746BE1A12944CD](https://github.com/iiasa/message_ix/assets/147787279/a911ee46-6c79-4144-86a7-f36086e41627)
<body>

growth_new_capacity_up | node_loc \| tec \| year_vtg
-- | --
soft_new_capacity_up


</body>
</html><html>
<body>

growth_new_capacity_up | node_loc \| tec \| year_vtg
-- | --
soft_new_capacity_up


</body>
</html>

**Reply by meng25meng** [reply]
> > 1.我的目标是得到全国在2030-2050年每个省份的装机和发电量情况。
> 
> 然后，我会建议联系那些在 MESSAGE China 工作了很长时间的研究人员。通过电力部门和省级建模来检查这一点。以及作者在相关著作中列出的其他研究人员（例如此处的次区域煤炭[）](https://doi.org/10.1029/2022EF003255)[，](https://doi.org/10.1021/acs.est.2c08171)[看看](https://www.sciencedirect.com/science/article/abs/pii/S0360544220327511?via%3Dihub)是否有更易于访问的基线，您可以将其用作构建场景的中间研究输入。[](https://doi.org/10.1029/2022EF003255)[](https://www.sciencedirect.com/science/article/abs/pii/S0360544220327511?via%3Dihub)
> 
> > 5.在发电同行技术中有别的能源都有发电：在这里水平和边际分别代表什么？而且在边际中还会有EPS出现，
> 
> 我看到您添加的 hydro_* 在您的 gdx 输出中具有非零级别，这很好。您的模型构建良好且已解决。 请使用中的值（[在此处](https://www.gams.com/latest/docs/UG_Equations.html)`level`检查级别和边际的定义，还有 EPS）。[](https://www.gams.com/latest/docs/UG_Equations.html)
> 
> > 2. 乡村提示，我首先检查了我的excel层级，使用了GD省的能源结构，如下：
> 
> 它看起来不错（并且由于你有一些水力水平，所以它应该不错）。
> 
> > 4.在gdx文件中在全国层面的煤和铀开采对应的ACT都为0，这个是因为还没有产生结果吗？所以为0，还是别的原因，
> > 3.您的提示中对于资源的表示。我在Excel中输入如下：
> 
> 那么现在的问题是，您为其提供的资源输入技术显示零活动？那么您怀疑这些零活动是添加参数时的技术问题导致的，还是系统偏爱水力发电的结果？ 一种诊断方法是添加一种新的虚拟资源“水”，并将其提供给资源成本较高或较低的水力发电。
> 
> 但最终第一点应该能最直接地解决您的问题。

还有一点，关于化石燃料的燃料成本需要在ext-coal这里添加还是在sub-coal转化成电力这一级别呢？不发电是不是也与燃料成本有关呢？燃料成本包含在投资成本、运维成本和可变成本之中吗？

**Reply by meng25meng** [reply]
> > 1.我的目标是得到全国在2030-2050年每个省份的装机和发电量情况。
> 
> 然后，我会建议联系那些在 MESSAGE China 工作了很长时间的研究人员。通过电力部门和省级建模来检查这一点。以及作者在相关著作中列出的其他研究人员（例如此处的次区域煤炭[）](https://doi.org/10.1029/2022EF003255)[，](https://doi.org/10.1021/acs.est.2c08171)[看看](https://www.sciencedirect.com/science/article/abs/pii/S0360544220327511?via%3Dihub)是否有更易于访问的基线，您可以将其用作构建场景的中间研究输入。[](https://doi.org/10.1029/2022EF003255)[](https://www.sciencedirect.com/science/article/abs/pii/S0360544220327511?via%3Dihub)
> 
> > 5.在发电同行技术中有别的能源都有发电：在这里水平和边际分别代表什么？而且在边际中还会有EPS出现，
> 
> 我看到您添加的 hydro_* 在您的 gdx 输出中具有非零级别，这很好。您的模型构建良好且已解决。 请使用中的值（[在此处](https://www.gams.com/latest/docs/UG_Equations.html)`level`检查级别和边际的定义，还有 EPS）。[](https://www.gams.com/latest/docs/UG_Equations.html)
> 
> > 2. 乡村提示，我首先检查了我的excel层级，使用了GD省的能源结构，如下：
> 
> 它看起来不错（并且由于你有一些水力水平，所以它应该不错）。
> 
> > 4.在gdx文件中在全国层面的煤和铀开采对应的ACT都为0，这个是因为还没有产生结果吗？所以为0，还是别的原因，
> > 3.您的提示中对于资源的表示。我在Excel中输入如下：
> 
> 那么现在的问题是，您为其提供的资源输入技术显示零活动？那么您怀疑这些零活动是添加参数时的技术问题导致的，还是系统偏爱水力发电的结果？ 一种诊断方法是添加一种新的虚拟资源“水”，并将其提供给资源成本较高或较低的水力发电。
> 
> 但最终第一点应该能最直接地解决您的问题。

您好，我貌似找到了答案，谢谢诶您之前的提醒，这个与input有关，我把input的值清空了，最后电量部分就又有发电量了。我有个疑问就是
1.这个input的单位unit是自己可以更改吗？
2.有没有一种约束可以让每个省份所有能源都发电？因为最小成本的的话会把成本高的技术就不发电了，这样不是太合理。

**Comment by yiyi1991** [comment]
>还有一点，关于化石燃料的燃料成本需要在ext-coal这里添加还是在sub-coal转化成电力这一级别呢？不发电是不是也与燃料成本有关呢？燃料成本包含在投资成本、运维成本和可变成本之中吗？

Can you give the definition of your "燃料成本"?
If you mean the price of the commodity coal, it does not need to be counted as an additional var_cost. 
(In your application, please also pay attention to sub-regional energy trade and international energy trade.)

**Reply by meng25meng** [reply]
> > 还有一些一点，关于化石燃料的燃料成本赞成ext-coal这里添加还是在sub-coal转化成电力级别呢？不发电是不是也与燃料成本有关呢？燃料成本包含在投资成本、运维成本和可变成本之中吗？
> 
> 你能给出“燃料成本”的定义吗？ 如果你指的是商品煤的价格，它不需要算作额外的 var_cost。 （在你的应用中，还请关注次区域能源贸易和国际能源贸易。）

由于内存限制，我还是没加上开采技术这一层级，在煤电技术那里直接转化成电量，能运算出来，但是为什么他每种能源在每一年的装机容量都相同，而且有的技术装机容量是0

**Comment by yiyi1991** [comment]
> 您好，我貌似找到了答案，谢谢诶您之前的提醒，这个与input有关，我把input的值清空了，最后电量部分就又有发电量了。

Great.
 
>我有个疑问就是, 1.这个input的单位unit是自己可以更改吗？

I think yes, if you take care of the corresponding resource costs, keep the single technology aligned with other technologies, and change the corresponding unit in reporting. I have not tested it on the supply-side technologies by myself but at the end the optimization happens at the activity level (the levelized system cost per unit of activity).  

> 2.有没有一种约束可以让每个省份所有能源都发电？因为最小成本的的话会把成本高的技术就不发电了，这样不是太合理。

Minimizing costs is the core of this optimization model. Of course, one can set the `bound_activity_lo` for specific regions.

**Comment by yiyi1991** [comment]
> 但是为什么他每种能源在每一年的装机容量都相同

I guess you misread the model year and the vintage year? 
If not, have you set any constraints on `growth_new_capacity_up` (which gives an upper limit on capacity addition) or `bound_total_capacity_lo` (which gives a lower limit on total capacity)?

**Reply by meng25meng** [reply]
> > 但是为什么他演示了在每一年的装机容量都相同
> 
> 我猜你看错了车型年份和年份？ 如果没有，你有没有设置任何限制`growth_new_capacity_up`（这给出了容量增加的上限）或`bound_total_capacity_lo`（这给出了总容量的下限）？

所有能源我都设置了三种约束，bound-new-capacity-up，bound-total-capacity-up/lo这三种，最后结果装机一直保持在bound-total-capacity-up这个边界上，没有变化。

**Reply by meng25meng** [reply]
> > 但是为什么他每种能源在每一年的装机容量都相同
> 
> I guess you misread the model year and the vintage year? If not, have you set any constraints on `growth_new_capacity_up` (which gives an upper limit on capacity addition) or `bound_total_capacity_lo` (which gives a lower limit on total capacity)?

![image](https://github.com/iiasa/message_ix/assets/147787279/d5859950-ec17-4132-97d7-51066640a4ae)
在excel中我只添加了capacity的相关约束，没有activity的约束，如果要加actinity的上线限约束是否需要先添加historit-activity的数据呢？在之前的Excel中没有加activity的数据

**Comment by yiyi1991** [comment]
> 所有能源我都设置了三种约束，bound-new-capacity-up，bound-total-capacity-up/lo这三种，最后结果装机一直保持在bound-total-capacity-up这个边界上，没有变化。

Yes, `bound_new_capacity_up` keeps new capacity in that period no more than a specific level. 
Similarly, `bound_total_capacity_up` keeps total capacity in that period no more than a specific level. 

See core formulation [here](https://docs.messageix.org/en/latest/model/MESSAGE/model_core.html#equation-new-capacity-constraint-lo).

**Reply by meng25meng** [reply]
> > 所有能源我都设置了三种约束，bound-new-capacity-up，bound-total-capacity-up/lo这三种，最后结果装机一直保持在bound-total-capacity-up这个边界上，没有变化。
> 
> 是的，`bound_new_capacity_up`将这段时间内的新增产能控制在一定水平之内。 同样，`bound_total_capacity_up`将这段时间内的总产能控制在一定水平之内。
> 
> [请参阅此处的](https://docs.messageix.org/en/latest/model/MESSAGE/model_core.html#equation-new-capacity-constraint-lo)核心公式。

您说很对，目前就是各能源装机情况都符合他的约束上下限，现在有两个问题不清楚该如何解决，1.如何让每种能源每个时间节点都有增长，我发现有的能源装机一直不变，不符合实际
2.一个省份内有的能源装机变成了0，一直都没有装机，
这两个问题感觉调试起来有点麻烦，您能提供一下思路吗？非常感谢。或者需要再添加一下什么约束吗？

**Comment by yiyi1991** [comment]
> 我发现有的能源装机一直不变

You mean that you observed the capacity variables kept at a specific level, which does NOT equal the up (or low) limit you set for their boundaries? (To me this 15.25 looks like the system prefers the cheapest hydropower and used up all its available capacity in each period, which is the bound_total_capacity_up you set for them.)

Please check the differences of bound_new_capacity_up/lo, bound_total_capacity_up/lo, growth_new_capacity_up/lo. Hope that can help. 

Again, please consider contacting the researchers I mentioned here previously. 

> I will then suggest getting in contact with researchers who have been working on MESSAGE China for a longer time. Check [this](https://doi.org/10.1021/acs.est.2c08171) with power sectors and [this](https://doi.org/10.1029/2022EF003255) with province-level modeling. And other researchers in the author lists in related works (e.g., sub-regional coal [here](https://www.sciencedirect.com/science/article/abs/pii/S0360544220327511?via%3Dihub)), to see if there is a more accessible baseline you can use as an intermediate research input to build your scenarios on.

**Reply by meng25meng** [reply]
> You mean that you observed the capacity variables kept at a specific level, which does NOT equal the up (or low) limit you set for their boundaries? (To me this 15.25 looks like the system prefers the cheapest hydropower and used up all its available capacity in each period, which is the bound_total_capacity_up you set for them.)

  @yiyi1991 可能还需要您的帮助，模型在求解的时候报错为如下：这个我查找了网站上得关于这个commodity-balance的公式，发现还是找不出原因，找不到该如何去修改我的数据，报错如下：
Infeasibility row 'COMMODITY_BALANCE_GT(JL.electricity.final.2025.year)':  0 >= 13.086.
Presolve time = 0.02 sec. (10.31 ticks)
LP status(3): infeasible
Cplex Time: 0.02sec (det. 10.39 ticks)
Rerunning with presolve turned off and solving with LPMETHOD=0
CPXPARAM_Advance                                 2
CPXPARAM_Preprocessing_Presolve                  0
CPXPARAM_LPMethod                                1
CPXPARAM_Threads                                 4
CPXPARAM_Parallel                                1
CPXPARAM_Barrier_Limits_Iteration                100000000
CPXPARAM_TimeLimit                               1000000
CPXPARAM_Tune_TimeLimit                          200000

**Comment by yiyi1991** [comment]
Based on this log I see the model failed to reach the desirable 13.086 level for the commodity `JL.electricity.final` consumption in 2025, thus causing infeasibility. 

Are there any additional constraints that stay in the way of the technologies that consume the commodity `JL.electricity.final`?
Ore are there any additional constraints that stay in the way of the technologies that provide the commodity `JL.electricity.final`?
What is your first model year? 

(For the former, maybe there is a `growth_activity_up` constraint on technologies that consume the commodity electricity in 2020? Or maybe these technologies consume the commodity electricity but not necessarily the `JL.electricity`? Do you see the commodity electricity being used in other regions?)
